File: ml/features.py
"""
Cyber Chaukidaar - Seismic feature extraction (reference implementation, v2).

This is the *reference* definition of the 15 features used by the vibration
classifier.  The same maths is implemented in:
  - firmware/src/dsp/features.cpp   (ESP32, C++)
  - server/src/ml/features.js       (Node.js)

Keep all three in sync.  Feature order is part of the model contract.

Input: a window of band-passed vibration samples (g units) sampled at FS Hz.
Output: dict of features.  Pure Python, no numpy required.
"""

import math

FS = 100.0            # ADXL345 output data rate used by the nodes
WINDOW = 128          # samples per analysis window (1.28 s)

FEATURE_ORDER = [
    "rms",
    "peak",
    "peakToPeak",
    "variance",
    "dominantFrequency",
    "spectralEnergy",
    "spectralCentroid",
    "interPeakInterval",
    "zeroCrossingRate",
    "crestFactor",
    # v2 additions
    "kurtosis",          # impulsiveness: footsteps are spiky (high), engines/wind are not
    "spectralFlatness",  # 0 = tonal (engine), 1 = white noise (wind / rain)
    "lowBandRatio",      # share of spectral energy in 1-8 Hz
    "highBandRatio",     # share of spectral energy in 20-50 Hz
    "cadenceStrength",   # autocorrelation of the envelope at walking lags (0.3-0.9 s)
]


def _fft_magnitudes(signal):
    """Radix-2 FFT magnitudes for bins 0 .. N/2-1 (pure python)."""
    n = len(signal)
    size = 1
    while size < n:
        size <<= 1
    re = list(signal) + [0.0] * (size - n)
    im = [0.0] * size

    j = 0
    for i in range(1, size):
        bit = size >> 1
        while j & bit:
            j ^= bit
            bit >>= 1
        j |= bit
        if i < j:
            re[i], re[j] = re[j], re[i]
            im[i], im[j] = im[j], im[i]

    length = 2
    while length <= size:
        ang = -2.0 * math.pi / length
        wr, wi = math.cos(ang), math.sin(ang)
        for start in range(0, size, length):
            cr, ci = 1.0, 0.0
            half = length // 2
            for k in range(half):
                a = start + k
                b = a + half
                tr = re[b] * cr - im[b] * ci
                ti = re[b] * ci + im[b] * cr
                re[b] = re[a] - tr
                im[b] = im[a] - ti
                re[a] += tr
                im[a] += ti
                cr, ci = cr * wr - ci * wi, cr * wi + ci * wr
        length <<= 1

    half = size // 2
    mags = [math.sqrt(re[k] ** 2 + im[k] ** 2) / size for k in range(half)]
    return mags, size


def extract(signal, fs=FS):
    n = len(signal)
    if n < 8:
        return {k: 0.0 for k in FEATURE_ORDER}

    mean = sum(signal) / n
    sig = [v - mean for v in signal]

    sum_sq = 0.0
    sum_4 = 0.0
    mn = sig[0]
    mx = sig[0]
    max_abs = 0.0
    for v in sig:
        v2 = v * v
        sum_sq += v2
        sum_4 += v2 * v2
        if v < mn:
            mn = v
        if v > mx:
            mx = v
        a = abs(v)
        if a > max_abs:
            max_abs = a

    rms = math.sqrt(sum_sq / n)
    variance = sum_sq / n
    peak = max_abs
    p2p = mx - mn
    crest = peak / rms if rms > 1e-6 else 0.0
    kurtosis = (sum_4 / n) / (variance * variance) if variance > 1e-12 else 0.0
    if kurtosis > 50.0:
        kurtosis = 50.0

    zc = 0
    for i in range(1, n):
        if (sig[i - 1] < 0 <= sig[i]) or (sig[i - 1] >= 0 > sig[i]):
            zc += 1
    zcr = zc / (n / fs)

    # spectral
    mags, size = _fft_magnitudes(sig)
    dom_freq = 0.0
    max_mag = -1.0
    energy = 0.0
    wsum = 0.0
    msum = 0.0
    low = 0.0
    high = 0.0
    log_sum = 0.0
    bins = 0
    for k in range(1, len(mags)):
        m = mags[k]
        f = k * fs / size
        e = m * m
        energy += e
        wsum += f * m
        msum += m
        if 1.0 <= f <= 8.0:
            low += e
        if 20.0 <= f <= 50.0:
            high += e
        log_sum += math.log(e + 1e-12)
        bins += 1
        if m > max_mag:
            max_mag = m
            dom_freq = f
    centroid = wsum / msum if msum > 0 else 0.0
    low_ratio = low / energy if energy > 1e-12 else 0.0
    high_ratio = high / energy if energy > 1e-12 else 0.0
    arith = energy / bins if bins else 0.0
    flatness = math.exp(log_sum / bins) / arith if bins and arith > 1e-12 else 0.0

    # inter-peak interval (ms) between prominent peaks with a 60 ms refractory
    thr = max_abs * 0.45
    refractory = int(0.06 * fs)
    last = -10_000
    intervals = []
    for i in range(1, n - 1):
        if sig[i] > thr and sig[i] > sig[i - 1] and sig[i] >= sig[i + 1]:
            if i - last >= refractory:
                if last >= 0:
                    intervals.append((i - last) / fs * 1000.0)
                last = i
    ipi = sum(intervals) / len(intervals) if intervals else 0.0

    # cadence strength: normalised autocorrelation of the smoothed |signal| envelope
    env = []
    for i in range(n):
        lo = max(0, i - 2)
        hi = min(n, i + 3)
        env.append(sum(abs(sig[j]) for j in range(lo, hi)) / (hi - lo))
    emean = sum(env) / n
    env = [v - emean for v in env]
    e0 = sum(v * v for v in env)
    cadence = 0.0
    if e0 > 1e-12:
        lag_min = int(0.3 * fs)
        lag_max = min(int(0.9 * fs), n - 8)
        for lag in range(lag_min, lag_max + 1):
            s = 0.0
            for i in range(n - lag):
                s += env[i] * env[i + lag]
            r = s / e0
            if r > cadence:
                cadence = r
    if cadence > 1.0:
        cadence = 1.0

    return {
        "rms": rms,
        "peak": peak,
        "peakToPeak": p2p,
        "variance": variance,
        "dominantFrequency": dom_freq,
        "spectralEnergy": energy,
        "spectralCentroid": centroid,
        "interPeakInterval": ipi,
        "zeroCrossingRate": zcr,
        "crestFactor": crest,
        "kurtosis": kurtosis,
        "spectralFlatness": flatness,
        "lowBandRatio": low_ratio,
        "highBandRatio": high_ratio,
        "cadenceStrength": cadence,
    }


def as_vector(features):
    return [float(features[k]) for k in FEATURE_ORDER]

File: server/src/ml/features.js
'use strict';
/**
 * Seismic feature extractor v2 — Node.js port of ml/features.py.
 * MUST stay numerically aligned with ml/features.py and firmware/src/dsp/features.cpp.
 */

const FEATURE_ORDER = [
  'rms', 'peak', 'peakToPeak', 'variance', 'dominantFrequency',
  'spectralEnergy', 'spectralCentroid', 'interPeakInterval', 'zeroCrossingRate', 'crestFactor',
  'kurtosis', 'spectralFlatness', 'lowBandRatio', 'highBandRatio', 'cadenceStrength'
];

function fftMagnitudes(signal) {
  const n = signal.length;
  let size = 1;
  while (size < n) size <<= 1;
  const re = new Float64Array(size);
  const im = new Float64Array(size);
  for (let i = 0; i < n; i++) re[i] = signal[i];
  let j = 0;
  for (let i = 1; i < size; i++) {
    let bit = size >> 1;
    while (j & bit) { j ^= bit; bit >>= 1; }
    j |= bit;
    if (i < j) {
      let t = re[i]; re[i] = re[j]; re[j] = t;
      t = im[i]; im[i] = im[j]; im[j] = t;
    }
  }
  for (let len = 2; len <= size; len <<= 1) {
    const ang = -2 * Math.PI / len;
    const wr = Math.cos(ang), wi = Math.sin(ang);
    for (let start = 0; start < size; start += len) {
      let cr = 1, ci = 0;
      const half = len >> 1;
      for (let k = 0; k < half; k++) {
        const a = start + k, b = a + half;
        const tr = re[b] * cr - im[b] * ci;
        const ti = re[b] * ci + im[b] * cr;
        re[b] = re[a] - tr; im[b] = im[a] - ti;
        re[a] += tr; im[a] += ti;
        const ncr = cr * wr - ci * wi;
        ci = cr * wi + ci * wr;
        cr = ncr;
      }
    }
  }
  const half = size >> 1;
  const mags = new Float64Array(half);
  for (let k = 0; k < half; k++) mags[k] = Math.sqrt(re[k] * re[k] + im[k] * im[k]) / size;
  return { mags, size };
}

function zeroFeatures() {
  const f = {};
  for (const k of FEATURE_ORDER) f[k] = 0;
  return f;
}

/**
 * @param {number[]} signal  band-passed vibration magnitude (g), any length >= 8
 * @param {number} fs        sample rate Hz
 */
function extract(signal, fs = 100) {
  const n = signal.length;
  if (!n || n < 8) return zeroFeatures();
  let mean = 0;
  for (let i = 0; i < n; i++) mean += signal[i];
  mean /= n;

  const sig = new Float64Array(n);
  let sumSq = 0, sum4 = 0, mn = Infinity, mx = -Infinity, maxAbs = 0;
  for (let i = 0; i < n; i++) {
    const v = signal[i] - mean;
    sig[i] = v;
    const v2 = v * v;
    sumSq += v2; sum4 += v2 * v2;
    if (v < mn) mn = v;
    if (v > mx) mx = v;
    const a = Math.abs(v);
    if (a > maxAbs) maxAbs = a;
  }
  const rms = Math.sqrt(sumSq / n);
  const variance = sumSq / n;
  const crest = rms > 1e-6 ? maxAbs / rms : 0;
  let kurtosis = variance > 1e-12 ? (sum4 / n) / (variance * variance) : 0;
  if (kurtosis > 50) kurtosis = 50;

  let zc = 0;
  for (let i = 1; i < n; i++) {
    if ((sig[i - 1] < 0 && sig[i] >= 0) || (sig[i - 1] >= 0 && sig[i] < 0)) zc++;
  }
  const zcr = zc / (n / fs);

  const { mags, size } = fftMagnitudes(sig);
  let domFreq = 0, maxMag = -1, energy = 0, wsum = 0, msum = 0, low = 0, high = 0, logSum = 0, bins = 0;
  for (let k = 1; k < mags.length; k++) {
    const m = mags[k];
    const f = k * fs / size;
    const e = m * m;
    energy += e; wsum += f * m; msum += m;
    if (f >= 1 && f <= 8) low += e;
    if (f >= 20 && f <= 50) high += e;
    logSum += Math.log(e + 1e-12);
    bins++;
    if (m > maxMag) { maxMag = m; domFreq = f; }
  }
  const centroid = msum > 0 ? wsum / msum : 0;
  const lowRatio = energy > 1e-12 ? low / energy : 0;
  const highRatio = energy > 1e-12 ? high / energy : 0;
  const arith = bins ? energy / bins : 0;
  const flatness = bins && arith > 1e-12 ? Math.exp(logSum / bins) / arith : 0;

  const thr = maxAbs * 0.45;
  const refractory = Math.floor(0.06 * fs);
  let last = -10000, sumInt = 0, cnt = 0;
  for (let i = 1; i < n - 1; i++) {
    if (sig[i] > thr && sig[i] > sig[i - 1] && sig[i] >= sig[i + 1]) {
      if (i - last >= refractory) {
        if (last >= 0) { sumInt += (i - last) / fs * 1000; cnt++; }
        last = i;
      }
    }
  }
  const ipi = cnt ? sumInt / cnt : 0;

  // cadence strength: normalised autocorrelation of the smoothed |signal| envelope
  const env = new Float64Array(n);
  let emean = 0;
  for (let i = 0; i < n; i++) {
    const lo = Math.max(0, i - 2), hi = Math.min(n, i + 3);
    let s = 0;
    for (let j = lo; j < hi; j++) s += Math.abs(sig[j]);
    env[i] = s / (hi - lo);
    emean += env[i];
  }
  emean /= n;
  let e0 = 0;
  for (let i = 0; i < n; i++) { env[i] -= emean; e0 += env[i] * env[i]; }
  let cadence = 0;
  if (e0 > 1e-12) {
    const lagMin = Math.floor(0.3 * fs), lagMax = Math.min(Math.floor(0.9 * fs), n - 8);
    for (let lag = lagMin; lag <= lagMax; lag++) {
      let s = 0;
      for (let i = 0; i < n - lag; i++) s += env[i] * env[i + lag];
      const r = s / e0;
      if (r > cadence) cadence = r;
    }
  }
  if (cadence > 1) cadence = 1;

  return {
    rms, peak: maxAbs, peakToPeak: mx - mn, variance,
    dominantFrequency: domFreq, spectralEnergy: energy, spectralCentroid: centroid,
    interPeakInterval: ipi, zeroCrossingRate: zcr, crestFactor: crest,
    kurtosis, spectralFlatness: flatness, lowBandRatio: lowRatio, highBandRatio: highRatio, cadenceStrength: cadence
  };
}

/** Spectrum for the dashboard: energy in bins (Hz) up to Nyquist. */
function spectrum(signal, fs = 100, bins = 24) {
  const n = signal.length;
  if (n < 8) return [];
  let mean = 0;
  for (let i = 0; i < n; i++) mean += signal[i];
  mean /= n;
  const sig = signal.map(v => v - mean);
  const { mags, size } = fftMagnitudes(sig);
  const out = new Array(bins).fill(0);
  const nyq = fs / 2;
  for (let k = 1; k < mags.length; k++) {
    const f = k * fs / size;
    const b = Math.min(bins - 1, Math.floor(f / nyq * bins));
    out[b] += mags[k];
  }
  return out.map((v, i) => ({ hz: Math.round((i + 0.5) * nyq / bins), mag: Number(v.toFixed(5)) }));
}

/** Envelope + detected step peaks for the scope overlay. */
function envelopeAndPeaks(signal, fs = 100) {
  const n = signal.length;
  if (n < 8) return { env: [], peaks: [] };
  let mean = 0;
  for (let i = 0; i < n; i++) mean += signal[i];
  mean /= n;
  const env = new Array(n);
  for (let i = 0; i < n; i++) {
    const lo = Math.max(0, i - 3), hi = Math.min(n, i + 4);
    let s = 0;
    for (let j = lo; j < hi; j++) s += Math.abs(signal[j] - mean);
    env[i] = s / (hi - lo);
  }
  let maxAbs = 0;
  for (let i = 0; i < n; i++) maxAbs = Math.max(maxAbs, Math.abs(signal[i] - mean));
  const thr = Math.max(0.02, maxAbs * 0.45);
  const refractory = Math.floor(0.25 * fs);
  const peaks = [];
  let last = -10000;
  for (let i = 1; i < n - 1; i++) {
    const v = signal[i] - mean;
    if (v > thr && v > signal[i - 1] - mean && v >= signal[i + 1] - mean && i - last >= refractory) { peaks.push(i); last = i; }
  }
  return { env, peaks };
}

function toVector(features) {
  return FEATURE_ORDER.map(k => Number(features[k]) || 0);
}

module.exports = { FEATURE_ORDER, extract, spectrum, envelopeAndPeaks, toVector, fftMagnitudes };
